### PR TITLE
01 fulfillment base

### PR DESCRIPTION
**Pull request by sujithvn**
Signed-off-by: Sujith mail.sujithvn@gmail.com

---
# This PR has been replaced by PR #6610 
---

Resolves #6471 
Impact: **breaking**
Type: **feature**

## Issue

In the current system, the only way for order fulfillment is via shipping which is hardcoded into the system. We need to introduce the flexibility for users to add their own fulfillment types via plugins.

## Solution

We are creating a base fulfillment plugin which would enable other fulfillment types (like shipping, pickup, digital) to be introduced via plugins. Each of the newly introduced fulfillment type plugin would need to have the specific fulfillment methods also to be added as separate plugins. Example, fulfillment type 'pickup' could have fulfillment methods like 'store pickup' and 'curb-side pickup'.

## PR Details in order

```
feat/fulfillment-types
 | 
 | __ 01 fulfillment base <--- This PR #6570 
 |        | __ 01-01 fulfillment-type-shipping       #6579
 |        | __ 01-02 fulfillment-type-pickup. #6577 
 |__02 fulfillment-type-impacted-plugins. [carts, products, orders, catalogs] #6582 
 |__03 placeOrder refactor #6583 

```
## Existing PRs based on the #6480 (to be updated to point to #6570) 
We shall change the base branch of the 3 existing PRs (i18n #6545, dataMigration #6544, setDefaultFFtype #6543) to  point to 01 fulfillment base

##### API-PLUGIN-FULFILLMENT
This is the base plugin, basic features listed below. 
Mutations
* createFulfillmentType - this is for internal call only from ff-type plugins (NEW)
* createFulfillmentMethod  - this is for internal call only from ff-method plugins (NEW)
* updateFulfillmentType - public mutation which allows to update certain specific fields of the ff-type (NEW)
* updateFulfillmentMethod - public mutation which allows to update certain specific fields of the ff-type (NEW)
* updateFulfillmentOptionsForGroup - Existing code/functionality from the old shipment plugin. This updates the cart with quote details from all available ff-methods for the provided ff-type
* selectFulfillmentOptionForGroup - Existing code/functionality from the old shipment plugin. This updates the cart with one of the selected ff-method from the available quote presented by above mutation.

Queries
* getFulfillmentMethodsWithQuotes - Existing code/functionality from the old shipment plugin. Calls the quote providing functions by collecting them via functionByType. Calls are made only for the specific methods of the provided ff-type.
* getFulfillmentType - returns the ff-type matching the provided ID (NEW)
* getFulfillmentTypes - returns all the ff-types (NEW) 
* getFulfillmentMethods - returns all the ff-methods for the specified ff-type (NEW)

Others
* preStartup - extends all impacted schemas mostly with dynamic values for 'allowedValues' (NEW)
* startup - inserts the default 'undecided' entry to the new Fulfillment collection (NEW)
* registration - collects all the "registeredFulfillmentType" from ff-type plugins and adds allRegisteredFulfillmentTypes to context (contextAddition) (NEW)
* util folder - copied over from shipment plugin (Existing code)

## Breaking changes

Since this is the new way of handling fulfillment, the existing shipment plugins would not work along with this. User would need to migrate the data (migration script will be developed) and remove the shipment plugins (api-plugin-shipments and api-plugin-shipments-flat-rate).

## Testing
All this is new code and at this stage this is not integrated to reaction, meaning the plugins.json is not updated to include this plugin. Hence proper testing can be carried out only after merging in the remaining PRs.

### CODE REVIEW

File: packages/api-plugin-fulfillment/README.md
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/README.md:1]: Please remove all this boilerplate and include complete documentation on what this plugin provides.

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js]: At some point we will go back to generating documentation from JSDocs so I don't think "see SimpleSchema" is going to cut it.

File: packages/api-plugin-fulfillment/README.md
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/README.md:1]: Please remove this boilerplate and replace with documentation

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js]: Schemas should normally start with a capital letter

File: packages/api-plugin-fulfillment/src/mutations/updateFulfillmentOptionsForGroup.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/updateFulfillmentOptionsForGroup.js]: Don't really know what format these JSDocs are in, but they are not consistent with the JSDocs we do everywhere else

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js]: filename is incorrect

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js]: I will once again comment that this string concatenation method feels like a hack

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentType.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/queries/getFulfillmentType.js]: `findOne` does not return a mongo cursor, it returns an object

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentTypes.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/queries/getFulfillmentTypes.js]: You can combine these two destructuring into one like we do everywhere else

File: packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentMethods.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentMethods.js:20]: Should we be checking permissions in the resolver rather than in the method? Feels like we should? Can you check what we do elsewhere?

File: packages/api-plugin-fulfillment/src/startup.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/startup.js]: These 2 destructuring can be combined

File: packages/api-plugin-fulfillment/src/util/getCartById.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/util/getCartById.js]: Confused why this is here and not in api-carts?

**Review by brent-hoover — CHANGES_REQUESTED** [review]

File: packages/api-plugin-fulfillment/.gitignore
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/.gitignore]: I don't think we need to include these in every plugin.

File: packages/api-plugin-fulfillment/src/index.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/index.js]: If these are new indexes they don't need to include the `c2` which is a legacy Meteor thing that shouldn't be applied to new work

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js:41]: So if you have the _id record than isn't the rest of these query parameters unnecessary? Especially since you can't create two records with the same _id

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js]: I don't think there's any reason to do this in a brand new collection

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.test.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.test.js:21]: Just my opinion but I think using snapshots are fine for frontend testing but for backend PR's they make it unnecessarily brittle. Open to second opinions on that.

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js]: Does this update the selected fulfillment type or does it create a fulfillment type? Is this copy pasta?

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js]: Should combine these two destructuring statements into one

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js]: So do we allow two records with the same fulfillmentType but a different name?

File: packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js]: Is this for flat rate or is the summary incorrect?

File: packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js]: Combine these two destructing statements into one

File: packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js]: Not sure why this comment is here?

File: packages/api-plugin-fulfillment/src/preStartup.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/preStartup.js]: I think I would prefer rather than storing this directly on context that we namespaced it

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js
Comment (brent-hoover) [reply packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js]: > I will once again comment that this string concatenation method feels like a hack

Instead of doing this I wonder if we could store a map of functions with a "type" attribute that would allow us not to depend on string concatenation?

File: packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentTypes.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentTypes.js]: Is this correct? If we are just returning one object why are we using pagination?

File: packages/api-plugin-fulfillment/src/resolvers/index.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/resolvers/index.js]: Is there a reason this comment is just hanging out here by itself?

File: packages/api-plugin-fulfillment/src/schemas/schema.graphql
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/schemas/schema.graphql:197]: Should this be required?

File: packages/api-plugin-fulfillment/src/simpleSchemas.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/simpleSchemas.js:10]: Why do we have a JSDoc here and then for no other Schemas?

File: packages/api-plugin-fulfillment/src/simpleSchemas.js
Comment (brent-hoover) [inline packages/api-plugin-fulfillment/src/simpleSchemas.js]: Probably should start with a capital letter

**Review by brent-hoover — CHANGES_REQUESTED** [review]
Just a few more comments

File: packages/api-plugin-fulfillment/.gitignore
Comment (sujithvn) [reply packages/api-plugin-fulfillment/.gitignore]: removed

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js]: Yes, as discussed, I have tested a solution as per the suggestion from @vanpho93 and it works.I will be implementing it once all the PRs are merged

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js]: Updated param fields

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js]: Fixed

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js]: fixed

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentType.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/queries/getFulfillmentType.js]: Fixed

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentTypes.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/queries/getFulfillmentTypes.js]: done

File: packages/api-plugin-fulfillment/src/startup.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/startup.js]: done

File: packages/api-plugin-fulfillment/.gitignore
Comment (sujithvn) [reply packages/api-plugin-fulfillment/.gitignore]: removed

File: packages/api-plugin-fulfillment/README.md
Comment (sujithvn) [reply packages/api-plugin-fulfillment/README.md:1]: Updated doc

Comment (sujithvn) [reply packages/api-plugin-fulfillment/README.md:1]: Updated doc

File: packages/api-plugin-fulfillment/src/mutations/updateFulfillmentOptionsForGroup.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/updateFulfillmentOptionsForGroup.js]: It was copied over from old shipment plugin. Fixed it.

File: packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentMethods.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentMethods.js:20]: I assume you're referring to `context.validatePermissions` and I see most of these calls happening in mutation/query methods

File: packages/api-plugin-fulfillment/src/util/getCartById.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/util/getCartById.js]: It was copied over from the old shipments plugin. As discussed, we have the same function inside api-carts also. So I shall make a new query for the one in api-cart and remove this one

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js:41]: The _id only gets us the ff-type. Along with the ff-type, I am querying for the specific ff-method also before adding a new ff-method.

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js]: removed the commented section

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js]: It creates. Fixed the @summary

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js]: Done

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (brent-hoover) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js:41]: This answer doesn't actually help clarify anything. Whatever you query with that _id is always going to give you the exact same record.

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentType.js]: Thanks for catching this bug. Updated the findOne to `await Fulfillment.findOne({ shopId, fulfillmentType });` 
name field is user editable and should not have been included in the selector. We allow only one entry per fulfillment-type

File: packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js]: Incorrect summary, fixed

File: packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js]: Done

File: packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/updateFulfillmentMethod.js]: removed (part of copy paste)

File: packages/api-plugin-fulfillment/src/resolvers/index.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/resolvers/index.js]: :-) 
No, it was copied over from api-shipments. Removed it.

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js:41]: Yes, we get the exact same ff-type record. But if the other parameters related to ff-method also match, then we need not proceed further since the new record is going to be a duplicate ff-method and we return error.
Else we should be further checking within the methods array of the returned record to see if we already have the new method

File: packages/api-plugin-fulfillment/src/simpleSchemas.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/simpleSchemas.js]: Updated

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js:41]: But I think we should remove the matching with name as you mentioned in a different review comment and retain only fulfillmentMethod matching. Else it would allow same method with different name

`methods: { $elemMatch: { fulfillmentMethod: method.fulfillmentMethod } }`

File: packages/api-plugin-fulfillment/src/schemas/schema.graphql
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/schemas/schema.graphql:197]: I am thinking the admin would be presented with all the existing values while trying to update. So the frontend could pass all the field values when calling this mutation.

File: packages/api-plugin-fulfillment/src/simpleSchemas.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/simpleSchemas.js:10]: Added JSDoc for the remaining Schemas as well

File: packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentTypes.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentTypes.js]: We are returning all the fulfillment types under the shop.
Fixed the 'return' text.

File: packages/api-plugin-fulfillment/src/util/getCartById.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/util/getCartById.js]: Since the changes will be spread over multiple PRs if implemented now, I shall wait for the currents PRs to be merged and create a new PR for this change.

File: packages/api-plugin-fulfillment/src/preStartup.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/preStartup.js]: Changed this to `context.fulfillment.registeredFulfillmentTypes`

File: packages/api-plugin-fulfillment/src/index.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/index.js]: updated

File: packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/queries/getFulfillmentMethodsWithQuotes.js]: Implemented

File: packages/api-plugin-fulfillment/src/util/getCartById.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/util/getCartById.js]: Implemented this change.
added the api-cart copy to queries and deleted api-fulfillment/src/util version

File: packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentMethods.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/resolvers/Query/getFulfillmentMethods.js:20]: @zenweasel Please let me know if this can be considered resolved

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.js:41]: @zenweasel Please let me know if this can be considered resolved

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.test.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.test.js:21]: Since there are no other opinions yet, I shall move this comment to the new PR (#6610) and resolve this comment. Comment in new PR (https://github.com/reactioncommerce/reaction/pull/6610/files#r1011288255)

@zenweasel Please let me know if this can be considered resolved

File: packages/api-plugin-fulfillment/src/schemas/schema.graphql
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/schemas/schema.graphql:197]: @zenweasel Please let me know if this can be considered resolved

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.test.js
Comment (brent-hoover) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.test.js:21]: Since there are no disagreements then my original comment stands. Snapshots are brittle and a very jest specific thing and I don't think we should introduce any new ones

File: packages/api-plugin-fulfillment/src/schemas/schema.graphql
Comment (brent-hoover) [reply packages/api-plugin-fulfillment/src/schemas/schema.graphql:197]: Your comment doesn't really answer my question. Why wouldn't enabled be required? Just enforcing it though the front end CLI is bad bad practice

File: packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.test.js
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/mutations/createFulfillmentMethod.test.js:21]: @zenweasel 
Sure. I have already copied over this comment to the new PR and shall make the changes in that branch. I will be resolving this comment.

File: packages/api-plugin-fulfillment/src/schemas/schema.graphql
Comment (sujithvn) [reply packages/api-plugin-fulfillment/src/schemas/schema.graphql:197]: Yes, that is right, I misunderstood.
Added the comment (https://github.com/reactioncommerce/reaction/pull/6610/files#r1011326448) in the PR #6610 and I'll be making this change there.
Resolving this comment.

**Comment by sujithvn** [comment]
This PR is closed and has been replaced by PR #6610

**Pull request closed without merging**
